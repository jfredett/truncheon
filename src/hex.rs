#![allow(dead_code)]

use std::{collections::HashMap, marker::ConstParamTy};

#[derive(Debug, PartialEq)]
struct AxialCoord {
    q: isize,
    r: isize
}

struct CubicCoord {

}

/// Maps a coordinate to a index counting from the origin provided
struct IndexCoord<const ORIGIN: Origin, const DIRECTION: Direction> {

}

#[derive(ConstParamTy, PartialEq, Eq)]
#[allow(non_camel_case_types)]
pub enum Direction {
    LR_TB,
    RL_TB,
    LR_BT,
    RL_BT
}

#[derive(ConstParamTy, PartialEq, Eq)]
pub enum Origin {
    TopLeft,
    TopRight,
    BottomLeft,
    BottomRight
}

#[derive(ConstParamTy, PartialEq, Eq)]
pub enum OffsetType {
    OddQ,
    OddR,
    EvenQ,
    EvenR
}

pub struct OffsetCoord<const TYPE: OffsetType> {
    x: isize,
    y: isize
}

///
/// ```
/// assert!(true);
///
/// ```
#[derive(Debug)]
pub struct Hexfield<const WIDTH: usize, const HEIGHT: usize, const ORIGIN: Origin, const DIRECTION: Direction, T> where T : Clone {
    contents: HashMap<AxialCoord, T>
}

impl<const WIDTH: usize, const HEIGHT: usize, const ORIGIN: Origin, const DIRECTION: Direction, T>
    Hexfield<WIDTH, HEIGHT, ORIGIN, DIRECTION, T>
where T : Clone {

    pub fn new() -> Self {
        Hexfield {
            contents: HashMap::new()
        }
    }
}


pub enum HexMovement {
    N,
    NE,
    NW,
    S,
    SE,
    SW
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    fn it_works() {
        let field : Hexfield<10, 10, { Origin::TopLeft }, { Direction::LR_TB }, ()> = Hexfield::new();

    }

}

/*

The goal of this thing is to reduce as much friction in the crawling process as possible.

Ideally the result is a CLI which talks to Foundry (or a WASM module that I load into Foundry)
which eventually generates a log of each step of the process, tracks time and date, allows
recording of events associated with that, etc.

A simple Ratatui UI which I punch the intended direction into, it should generate the various DCs according
to skill type based on the terrain and it's contents / tags etc.

The system should let me tag the resulting location, maybe record myself describing it, transcribe it, save it, etc.

I don't hate the raw data running to grist, but would prefer it eventually in foundry or in my own DB.

The thing should generate a couple maps -- the real map (where the players actually are), the implied map (where the players think they are), and ideally it should allow for a multi-planar system.

Coordinates should be endowed with an extra coordinate, "Plane" which cooresponds to:

1. Ground level, Earth
2. Sky, Earth
3. Underground, Earth
4. Ynn
5. The Stygian Library
6. etc.

Each plane may or may not be a hexcrawl, so in some planes I might switch to other methods of tracking, this tool is for that.

It should also, ideally, track depthcrawls, pointcrawls, etc. Something like (not to scale):

-------------------------------------------------
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
|-----------------------------------------------|
| Command line history                          |
|                                               |
|                                               |
|-----------------------------------------------|
| Command line                                  |
|-----------------------------------------------|


Maybe w/ the REST interface I could control Foundry _from_ this tool, which would be very cool indeed. If not, a websock into something custom could work.

ideally WFC happens here, for generating terrain automatically.

the ratatui maps will need a custom hex renderer, I think something like:


  ------
 /      \
/ A01234 \______
\SHRTMESG/      \
 \      /        \
  ------          -------
 /      \        /       \
/ A56789 \______/         \
\SHRTMESG/      \         /
 \      /        \       /
  ------          -------


A cell would be highlighted for party/player location on each screen, as well as actual location on each. Beneath the map might be a readout of the contents of the hex

Hexes can probably just be loaded/unloaded via Serde direct to files. a plane is just a folder of hexes, an unknown hex gets WFC'd.

Would be maybe cool to do a 'non-hex' system, where each hex is stored with adjaceny information, which would allow for a sort of natural non-euclidean,
non-uniform system that could contain both hex and depth crawls.

*/

