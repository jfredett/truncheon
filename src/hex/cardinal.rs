#[cfg(test)]
use derive_quickcheck_arbitrary::Arbitrary;

use crate::hex::coord::axial::AxialVector;


#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
#[cfg_attr(test, derive(Arbitrary))]
pub enum Cardinal {
    N,
    NE,
    NW,
    S,
    SE,
    SW
}

const S : Cardinal = Cardinal::S;
const SE : Cardinal = Cardinal::SE;
const SW : Cardinal = Cardinal::SW;
const N : Cardinal = Cardinal::N;
const NE : Cardinal = Cardinal::NE;
const NW : Cardinal = Cardinal::NW;

impl From<Cardinal> for AxialVector {
    fn from(value: Cardinal) -> Self {
        match value {
            Cardinal::N => { AxialVector::new(0,-1) },
            Cardinal::NE => { AxialVector::new(1,-1) },
            Cardinal::NW => { AxialVector::new(-1,0) },
            Cardinal::S => { AxialVector::new(0,1) },
            Cardinal::SE => { AxialVector::new(1,0) },
            Cardinal::SW => { AxialVector::new(-1,1) }
        }
    }
}

impl Cardinal {
    pub fn clockwise(&self) -> Self {
        match self {
            Cardinal::N => { NE },
            Cardinal::NE => { SE },
            Cardinal::SE => { S },
            Cardinal::S => { SW },
            Cardinal::SW => { NW },
            Cardinal::NW => { N },
        }
    }

    pub fn counterclockwise(&self) -> Self {
        match self {
            Cardinal::N => { NW },
            Cardinal::NW => { SW },
            Cardinal::SW => { S },
            Cardinal::S => { SE },
            Cardinal::SE => { NE },
            Cardinal::NE => { N },
        }
    }
}
