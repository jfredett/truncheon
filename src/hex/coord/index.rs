/// Maps a coordinate to a index counting from the origin provided
struct IndexCoord<const ORIGIN: axial::Point> {

}
