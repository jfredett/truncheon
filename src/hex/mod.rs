pub mod coord;
pub mod cardinal;
pub mod field;

/*
* Loose sources/notes


// for eventually calculating distance-to-horizon/number of revealed tiles

http://www.totally-cuckoo.com/distance_visible_to_the_horizon.htm

Height (ft)      Height (m)      Distance (miles)      Distance (km)
5                0.98            3.24                  5.21
10               3.05            4.18                  6.76
25               7.62            6.61                  10.62
50               15.24           9.35                  15.13
100              30.48           12.23                 19.63
200              60.96           18.72                 30.09
400              121.92          26.46                 42.65
1000             304.80          32.41                 52.14

The source of all hex magic:
https://www.redblobgames.com/grids/hexagons/

*/

// TODO:
// Something like a `World` struct which contains many `fields`, a `field` is a Hexfield, a
// Depthcrawl, a pointcrawl, etc. Fields contain `tiles` (`hex`es in the case of hexfield,
// locations for depth or pointcrawl), and `tiles` have some set of `connections`; for `hexes`
// there are 6 natural connections (to other hexes), but they might also connect to another field;
// or have one of their natural connections replaced with a connection to another field.
//
// On the `world` struct, there are some set of `cursor`s, which represent players, parties,
// enemies, or other mobile items.

/*

The goal of this thing is to reduce as much friction in the crawling process as possible.

Ideally the result is a CLI which talks to Foundry (or a WASM module that I load into Foundry)
which eventually generates a log of each step of the process, tracks time and date, allows
recording of events associated with that, etc.

A simple Ratatui UI which I punch the intended direction into, it should generate the various DCs according
to skill type based on the terrain and it's contents / tags etc.

The system should let me tag the resulting location, maybe record myself describing it, transcribe it, save it, etc.

I don't hate the raw data running to grist, but would prefer it eventually in foundry or in my own DB.

The thing should generate a couple maps -- the real map (where the players actually are), the implied map (where the players think they are), and ideally it should allow for a multi-planar system.

Coordinates should be endowed with an extra coordinate, "Plane" which cooresponds to:

1. Ground level, Earth
2. Sky, Earth
3. Underground, Earth
4. Ynn
5. The Stygian Library
6. etc.

Each plane may or may not be a hexcrawl, so in some planes I might switch to other methods of tracking, this tool is for that.

It should also, ideally, track depthcrawls, pointcrawls, etc. Something like (not to scale):

-------------------------------------------------
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
| NOTES  |  REAL MAP |  PLAYER MAP | TRAVEL LOG |
|-----------------------------------------------|
| Command line history                          |
|                                               |
|                                               |
|-----------------------------------------------|
| Command line                                  |
|-----------------------------------------------|


Maybe w/ the REST interface I could control Foundry _from_ this tool, which would be very cool indeed. If not, a websock into something custom could work.

ideally WFC happens here, for generating terrain automatically.

the ratatui maps will need a custom hex renderer, I think something like:


  ------
 /      \
/ A01234 \______
\SHRTMESG/      \
 \      /        \
  ------          -------
 /      \        /       \
/ A56789 \______/         \
\SHRTMESG/      \         /
 \      /        \       /
  ------          -------


A cell would be highlighted for party/player location on each screen, as well as actual location on each. Beneath the map might be a readout of the contents of the hex

Hexes can probably just be loaded/unloaded via Serde direct to files. a plane is just a folder of hexes, an unknown hex gets WFC'd.

Would be maybe cool to do a 'non-hex' system, where each hex is stored with adjaceny information, which would allow for a sort of natural non-euclidean,
non-uniform system that could contain both hex and depth crawls.

*/

