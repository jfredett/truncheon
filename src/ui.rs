pub async fn run() {
    todo!();
}
