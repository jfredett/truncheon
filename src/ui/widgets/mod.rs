pub mod canvas_placeholder;
pub mod hexmap;
pub mod placeholder;
pub mod svg;
