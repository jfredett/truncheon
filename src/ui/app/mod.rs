use ratatui::{crossterm::event::{Event, KeyCode}, layout::{Constraint, Layout, Rect}, style::{Color, Style}, Frame};
use truncheon::hex::{coord::axial, field::Field};

use std::{collections::HashMap, fmt::Debug, path::Path, str::FromStr};

use tui_logger::{LevelFilter, TuiLoggerLevelOutput, TuiLoggerSmartWidget, TuiWidgetState};

use lazy_static::lazy_static;

use crate::ui::{tui::Message, widgets::{hexmap::Hexmap, placeholder::Placeholder, svg::{SVGTemplate, SVG}}};

enum Mode {
    Insert,
    Command
}

pub struct UI {
    flags: HashMap<String, bool>,
    // layout: HashMap<String, Rect> // this gets async updated with the current layout whenever a resize event occurs.
    // UI
    mode: Mode,
    // NOTE: the update function is going to be used to render the SVG in the background, so I need
    // to remember the current Frame's area so I can grab the dimensions of the thing without
    // having the thing. This means that there may be a race when resizing, I think this could be
    // maintained via the resize event
    frame_area: Option<Rect>,
    layout: HashMap<String, Rect>,
    // State
    tuiloggerstate: TuiWidgetState,
    player_map: SVG,
}

impl Debug for UI {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Truncheon")
            .field("flags", &self.flags)
            .finish()
    }
}


impl UI {
    pub fn new() -> Self {
        Self {
            flags: HashMap::new(),
            mode: Mode::Command,
            tuiloggerstate: TuiWidgetState::new().set_default_display_level(LevelFilter::Trace),
            player_map: SVG::new(),
            // BUG: probably this should be set by something other than just the resize. Or maybe I
            // should just fire an initial resize somehow? results in a no-first-render bug.
            frame_area: None,
            layout: HashMap::new()
        }
    }

    // TODO: Pull this into some separate structure.
    // TODO: In addition, this should just specify what _should_ happen, not _do_ any of it, that's
    // for the Message to indicate. So message might be "Message:SwitchMode(Mode::Insert)" or
    // whatever.
    pub fn handle_events(&mut self, event: Event) -> Message {
        match event {
            Event::Resize(width, height) => { 
                tracing::info!("Caught resize old area: {:?}, new area {width},{height}", self.frame_area);
                let (x,y) = self.frame_area.map_or((0,0), |a| (a.x, a.y));
                self.frame_area = Some(Rect::new(x, y, width, height));
                self.build_layout();
            },
            Event::Key(key) => {
                match self.mode {
                    // Probably insert mode is just handled by the tile wholesale?
                    Mode::Insert => {
                        match key.code {
                            KeyCode::Esc => {
                                tracing::trace!("Switching to Command Mode");
                                self.mode = Mode::Command;
                            },
                            KeyCode::Char(c) => {
                                tracing::debug!("Recieved key: {:?}", c);
                            },
                            KeyCode::Backspace => {
                                // self.tile.handle_backspace();
                            },
                            KeyCode::Enter => {
                                // self.tile.handle_enter();
                            },
                            _ => {
                            }
                        }
                    },
                    // Command mode will eventually select the tile you want/start new tiles, etc.
                    Mode::Command => {
                        match key.code {
                            KeyCode::Char('i') => {
                                tracing::trace!("Switching to Insert Mode");
                                self.mode = Mode::Insert;
                            },
                            KeyCode::Char('q') => {
                                tracing::trace!("Quitting");
                                return Message::Quit;
                            },
                            // KeyCode::Down => {
                            // }
                            // KeyCode::Up => {
                            // }
                            _ => {
                            }
                        }
                    }
                }
            },
            _ => {}
            // Event::FocusGained => todo!(),
            // Event::FocusLost => todo!(),
            // Event::Mouse(_mouse_event) => todo!(),
            // Event::Paste(_) => todo!(),
        }

        Message::Noop
    }

    pub fn build_layout(&mut self) {
        if let Some(rect) = self.frame_area {
            tracing::info!("Rebuilding Layout");
            let chunks = PRIMARY_LAYOUT.split(rect);
            let upper_section = chunks[0];
            let log_section = chunks[1];
            let io_section = chunks[2];

            let chunks = UPPER_LAYOUT.split(upper_section);
            // Used to show the travel 'trail', both the intended and actual
            let trail_slice = chunks[0];
            // Used to show the intended location of the players -- NOTE: Should this be more like,
            // "svg_viewport_a" and have it's use determined by the SVGTemplate we pass it? That might
            // be better overall
            let player_map_slice = chunks[1];
            // Used to show the actual location of the players
            let gm_map_slice = chunks[2];


            let chunks = IO_SECTION_LAYOUT.split(io_section);
            let output_section = chunks[0];
            let input_section = chunks[1];

            self.layout = HashMap::from([
                ("trail_slice".to_string(), trail_slice),
                ("log_section".to_string(), log_section),
                ("gm_map_slice".to_string(), gm_map_slice),
                ("player_map_slice".to_string(), player_map_slice),
                ("output_section".to_string(), output_section),
                ("input_section".to_string(), input_section),
            ]);
        } else {
            tracing::debug!("Skipping layout since there is no frame yet");
        }
    }

    pub async fn update(&mut self) {
        if self.frame_area.is_none() { return; }
        // FIXME: This is non-ideal duplication.
        self.build_layout();
        // FIXME: This is also non-ideal duplication. I suppose I may want to swap out svgs during an
        // update, but this doesn't feel great as is
        self.player_map.update(self.layout["player_map_slice"], &mut SVGTemplate::from_file(Path::new("./tests/fixtures/svg/template.svg")));
    }

    fn tui_logger_widget(&self) -> TuiLoggerSmartWidget {
        TuiLoggerSmartWidget::default()
            .style_error(Style::default().fg(Color::Red))
            .style_debug(Style::default().fg(Color::Green))
            .style_warn(Style::default().fg(Color::Yellow))
            .style_trace(Style::default().fg(Color::Magenta))
            .style_info(Style::default().fg(Color::Cyan))
            .output_separator('|')
            .output_timestamp(Some("%H:%M:%S".to_string()))
            .output_level(Some(TuiLoggerLevelOutput::Long))
            .output_target(true)
            .output_file(true)
            .output_line(true)
            .state(&self.tuiloggerstate)
    }

    pub fn render(&self, frame: &mut Frame<'_>) {
        let tlw = self.tui_logger_widget();

        if self.frame_area.is_none() { return; }

        frame.render_widget(tlw, self.layout["log_section"]);

        frame.render_widget(&Placeholder::for_section(self.layout["trail_slice"]).text("TRAIL"), self.layout["trail_slice"]);
        frame.render_stateful_widget(&self.player_map, self.layout["player_map_slice"], &mut SVGTemplate::from_file(Path::new("./tests/fixtures/svg/template.svg")));
        frame.render_stateful_widget(Hexmap::default(), self.layout["gm_map_slice"], &mut Field::<isize>::new());
        frame.render_widget(&Placeholder::for_section(self.layout["output_section"]).text("OUTPUT"), self.layout["output_section"]);
        frame.render_widget(&Placeholder::for_section(self.layout["input_section"]).text("> INPUT"), self.layout["input_section"]);
    }
}

lazy_static! {
    static ref PRIMARY_LAYOUT : Layout = Layout::default()
        .direction(ratatui::layout::Direction::Vertical)
        .constraints([
                Constraint::Percentage(50),
                Constraint::Percentage(30),
                Constraint::Percentage(20),
                Constraint::Min(1),
            ].as_ref());

    static ref UPPER_LAYOUT : Layout = Layout::default()
        .direction(ratatui::layout::Direction::Horizontal)
        .constraints([
            Constraint::Percentage(10),
            Constraint::Percentage(45),
            Constraint::Percentage(45),
        ].as_ref());
    static ref IO_SECTION_LAYOUT : Layout = Layout::default()
        .direction(ratatui::layout::Direction::Vertical)
        .constraints([
            Constraint::Min(1),
            Constraint::Length(3)
        ].as_ref());

    static ref TEST_FIELD : Field<isize> = {
        let f = Field::new();
        f.insert(axial::Point::from_str("[0,0]").unwrap(), 0);
        f.insert(axial::Point::from_str("[0,1]").unwrap(), 1);
        f.insert(axial::Point::from_str("[1,0]").unwrap(), 2);
        f.insert(axial::Point::from_str("[1,1]").unwrap(), 3);
        f.insert(axial::Point::from_str("[0,-1]").unwrap(), -1);
        f.insert(axial::Point::from_str("[-1,0]").unwrap(), -2);
        f.insert(axial::Point::from_str("[-1,-1]").unwrap(), -3);
        f
    };
}
