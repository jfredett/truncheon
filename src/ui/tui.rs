// Taken and adapted from the original https://github.com/d-holguin/async-ratatui/blob/main/src/lib.rs
// used under the offered MIT License, thus copyright by 2024 d.holguin. Thanks Daniel!


use ratatui::backend::CrosstermBackend;
use ratatui::crossterm::event::{
    DisableMouseCapture, EnableMouseCapture, Event,
};
use ratatui::crossterm::terminal::{EnterAlternateScreen, LeaveAlternateScreen};
use ratatui::layout::{Constraint, Direction, Layout};
use ratatui::text::Text;
use ratatui::widgets::{Block, Borders, Paragraph};
use ratatui::{crossterm, Terminal};
use ratatui_image::picker::Picker;
use tui_logger::{LevelFilter, TuiLoggerFile, TuiLoggerLevelOutput};
use std::env;
use std::sync::Arc;
use tokio::sync::RwLock;
use std::time::Duration;
use tokio::sync::mpsc;
use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
use tokio::time;

use crate::ui::app::UI;

pub type Error = Box<dyn std::error::Error>;
pub type Result<T> = std::result::Result<T, Error>;

#[derive(Clone, Debug)]
pub enum Message {
    Noop,
    Quit,
    Tick,
    Render
}

pub struct Tui {
    pub terminal: Terminal<CrosstermBackend<std::io::Stdout>>,
    pub frame_rate: f64,
    pub tick_rate: f64,
    pub event_tx: UnboundedSender<Message>,
    pub event_rx: UnboundedReceiver<Message>,
    // TODO: UI -> generic 'Engine' which takes the update/event/render, can be shared.
    pub app: Arc<RwLock<UI>>
}

#[derive(Clone, Debug)]
pub enum UpdateCommand {
    None,
    Quit,
}

impl Tui {
    pub fn new(app: Arc<RwLock<UI>>, frame_rate: f64, tick_rate: f64) -> Result<Self> {
        let terminal = Terminal::new(CrosstermBackend::new(std::io::stdout()))?;
        let (event_tx, event_rx) = mpsc::unbounded_channel();
        Ok(Self {
            terminal,
            frame_rate,
            tick_rate,
            event_tx,
            event_rx,
            app
        })
    }

    fn enter(&mut self) -> Result<()> {
        crossterm::terminal::enable_raw_mode()?;
        crossterm::execute!(std::io::stdout(), EnterAlternateScreen, EnableMouseCapture)?;
        Ok(())
    }

    pub fn exit(&mut self) -> Result<()> {
        if crossterm::terminal::is_raw_mode_enabled()? {
            self.terminal.flush()?;
            crossterm::execute!(std::io::stdout(), LeaveAlternateScreen, DisableMouseCapture)?;
            crossterm::terminal::disable_raw_mode()?;
            self.terminal.show_cursor()?;
            println!("Terminal exited.");
        }
        Ok(())
    }

    pub async fn run(&mut self) -> Result<()> {
        use tracing_subscriber::prelude::*;

        // Set up the Tracing layer
        tracing_subscriber::registry()
            .with(tui_logger::TuiTracingSubscriberLayer)
            .init();

        // Set the log files
        tui_logger::init_logger(LevelFilter::Trace)?;
        tui_logger::set_default_level(LevelFilter::Trace);
        // prepare the log directory and file.
        let mut dir = env::temp_dir();

        // TODO: Replace with something like "app.name()"
        dir.push("truncheon.log");
        let file_options = TuiLoggerFile::new(dir.to_str().unwrap())
            .output_level(Some(TuiLoggerLevelOutput::Abbreviated))
            .output_file(false)
            .output_separator(':');

        eprintln!("dir: {}", &dir.display());
        tui_logger::set_log_file(file_options);


        // TODO: Set target dynamically as well
        tracing::debug!(target:"truncheon::ui", "Logging to {}", dir.to_str().unwrap());

        // FIXME: A bunch of this should probably live in like, an `init` function on app.

        self.enter()?;
        tracing::debug!(target:"truncheon::ui", "Entered terminal");
        self.terminal.clear()?;

        let tick_rate = Duration::from_secs_f64(1.0 / self.tick_rate);
        let frame_rate = Duration::from_secs_f64(1.0 / self.frame_rate);
        let mut tick_interval = time::interval(tick_rate);
        let mut frame_interval = time::interval(frame_rate);

        loop {
            tokio::select! {
                _tick = tick_interval.tick() => {
                    tracing::trace!("tick");
                    if let Err(e) = self.event_tx.send(Message::Tick) {
                        return Err(format!("Failed to tick: {:?}", e).into());
                    }
                }
                _frame = frame_interval.tick() => {
                    tracing::trace!("frame");
                    if let Err(e) = self.event_tx.send(Message::Render) {
                        return Err(format!("Failed to render frame: {:?}", e).into());
                    }
                }
                Some(message) = self.event_rx.recv() => {
                    tracing::trace!("event");
                    match self.update(message).await? {
                        UpdateCommand::Quit => return {
                            self.exit()?;
                            Ok(())
                        },
                        UpdateCommand::None => continue,
                    }
                }
                Ok(ready) = tokio::task::spawn_blocking(|| crossterm::event::poll(Duration::from_millis(100))) => {
                    tracing::trace!("poll");
                    match ready {
                        Ok(true) => {
                            let event = crossterm::event::read()?;
                            if let Err(e) = self.handle_event(event).await {
                                return Err(format!("Failed to handle event: {:?}", e).into());
                            }
                        }
                        Ok(false) => continue,
                        Err(e) => {
                            return Err(format!("Failed to poll for events: {:?}", e).into());
                        }
                    }
                }
            }
        }
    }

    // TODO: Inline, make the app#handle_events return a result
    async fn handle_event(&self, event: Event) -> Result<()> {
        let mut app = self.app.write().await;
        let response = app.handle_events(event);
        self.event_tx.send(response)?;
        Ok(())
    }

    async fn update(&mut self, message: Message) -> Result<UpdateCommand> {
        match message {
            Message::Noop => {},
            Message::Quit => { return Ok(UpdateCommand::Quit) },
            Message::Tick => {
                tracing::info!("Updating state");
                let mut guard = self.app.write().await;
                guard.update().await;
                drop(guard);
            },
            Message::Render => {
                tracing::info!("Rendering frame");
                self.view().await?;
            },
        }

        Ok(UpdateCommand::None)
    }

    async fn view(&mut self) -> Result<()> {
        // let counter : isize = std::random::random();
        // self.terminal.draw(|f| {
        //     let size = f.size();
        //     let chunks = Layout::default()
        //         .direction(Direction::Vertical)
        //         .constraints([Constraint::Min(1)].as_ref())
        //         .split(size);

        //     let block = Block::default().title("Hello").borders(Borders::ALL);
        //     let paragraph = Paragraph::new(Text::raw(format!("This is a test {}", counter))).block(block);
        //     f.render_widget(paragraph, chunks[0]);
        // })?;
        // Ok(())


        // BUG: This picker shit is killing the party.
        let picker = if cfg!(test) {
            // avoids an issue during testing by fixing the fontsize, normally this is unset for
            // the test
            Picker::from_fontsize((7, 12))
        } else {
            Picker::from_query_stdio().unwrap_or(Picker::from_fontsize((8,12)))
        };

        let shape = FrameShape {
            terminal_size: self.terminal.size().map(|s| (s.width, s.height))?,
            font_size: picker.font_size()
        };

        let app = self.app.read().await;
        self.terminal.draw(|f| {
            app.render(f, &shape)
        })?;

        Ok(())
    }
}

pub struct FrameShape {
    pub terminal_size: (u16, u16),
    pub font_size: (u16, u16)
}

impl Drop for Tui {
    fn drop(&mut self) {
        self.exit().expect("Failed to end terminal mode");
    }
}

