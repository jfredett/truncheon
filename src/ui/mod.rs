use std::sync::Arc;
use tokio::sync::RwLock;
use std::error::Error;

mod app;
mod widgets;
mod tui;

use app::UI;

use crate::ui::tui::Tui;


pub async fn run() -> Result<(), Box<dyn Error>> {
    let app = Arc::new(RwLock::new(UI::new()));
    let mut tui = Tui::new(app.clone(), 60.0, 100.0).unwrap();
    tui.run().await?;

    Ok(())
}

// async fn run_app<B: Backend>(terminal: &mut Terminal<B>, app: &mut Arc<RwLock<UI>>) -> io::Result<bool> {
//     use tracing_subscriber::prelude::*;

//     // Set up the Tracing layer
//     tracing_subscriber::registry()
//         .with(tui_logger::TuiTracingSubscriberLayer)
//         .init();

//     // Initialize the tui-logger widget
//     let _ = init_logger(LevelFilter::Trace);
//     set_default_level(LevelFilter::Trace);


// //     tracing::debug!(target:"truncheon::ui", "Logging to {}", dir.to_str().unwrap());
// //     tracing::debug!(target:"truncheon::ui", "Logging initialized");




//     // we are not okay, this method is bad.
//     return Ok(false);
// }
