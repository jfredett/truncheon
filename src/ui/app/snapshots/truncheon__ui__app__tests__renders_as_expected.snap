---
source: src/ui/app/mod.rs
expression: t.backend().buffer().clone()
---
Buffer {
    area: Rect { x: 0, y: 0, width: 64, height: 32 },
    content: [
        "┌────┐┌Hexmap─────────────────────┐┌Hexmap─────────────────────┐",
        "│TRAI││                           ││                           │",
        "│  L ││                           ││                           │",
        "│TRAI││                           ││                           │",
        "│  L ││                           ││                           │",
        "│TRAI││                           ││                           │",
        "│  L ││           ⣀⡀⠲⡀⡀           ││           ⣀⡀⠲⡀⡀           │",
        "│TRAI││          ⠰⣀⡀⠒⠁⡱           ││          ⠰⣀⡀⠒⠁⡱           │",
        "│  L ││          ⠰⣁⡱⠰⣁⡱           ││          ⠰⣁⡱⠰⣁⡱           │",
        "│TRAI││            ⠑⠒⠁            ││            ⠑⠒⠁            │",
        "│  L ││                           ││                           │",
        "│TRAI││                           ││                           │",
        "│  L ││                           ││                           │",
        "│TRAI││                           ││                           │",
        "│  L ││                           ││                           │",
        "└────┘└───────────────────────────┘└───────────────────────────┘",
        "┌Tui Ta┐┌Tui Log [log=0.0/s]───────────────────────────────────┐",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "└──────┘└──────────────────────────────────────────────────────┘",
        "┌──────────────────────────────────────────────────────────────┐",
        "│                            OUTPUT                            │",
        "│                            OUTPUT                            │",
        "│                            OUTPUT                            │",
        "└──────────────────────────────────────────────────────────────┘",
        "┌──────────────────────────────────────────────────────────────┐",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 6, fg: Red, bg: Reset, underline: Reset, modifier: NONE,
        x: 23, y: 6, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 47, y: 6, fg: Red, bg: Reset, underline: Reset, modifier: NONE,
        x: 52, y: 6, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 17, y: 7, fg: Red, bg: Reset, underline: Reset, modifier: NONE,
        x: 23, y: 7, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 46, y: 7, fg: Red, bg: Reset, underline: Reset, modifier: NONE,
        x: 52, y: 7, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 17, y: 8, fg: Red, bg: Reset, underline: Reset, modifier: NONE,
        x: 23, y: 8, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 46, y: 8, fg: Red, bg: Reset, underline: Reset, modifier: NONE,
        x: 52, y: 8, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 19, y: 9, fg: Red, bg: Reset, underline: Reset, modifier: NONE,
        x: 22, y: 9, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 48, y: 9, fg: Red, bg: Reset, underline: Reset, modifier: NONE,
        x: 51, y: 9, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
